### PR TITLE
geo/geomfn: implement ST_Reverse

### LINKED ISSUE #49018: geo/geomfn: implement ST_Reverse({geometry})
Implement `ST_Reverse` on arguments {geometry}, which should adopt [PostGIS behaviour](https://postgis.net/docs/ST_Reverse.html).

_Observers: Please react to this issue if you need this functionality._

For Geometry builtins, please do the following:
* Ideally add a relevant helper function in [`pkg/geo/geomfn`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo/geomfn) (parse and output related functions can go in [`pkg/geo`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo)). Add exhaustive unit tests here - you can run through example test cases and make sure that PostGIS and CRDB return the same result within a degree of accuracy (1cm for geography).
  * When using GEOS, you can reference the [C API](https://github.com/libgeos/geos/blob/master/capi/geos_c.h.in) for which functions are available. Unfortunately, Windows is not currently supported when using GEOS.
* Create a new builtin that references this function in [`pkg/sql/sem/builtins/geo_builtins.go`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/geo_builtins.go). Note that we currently do not support optional arguments, so we define functions that have optional arguments once without the optional argument (using the default value in the optional position), and once with the optional argument.
* Modify the tests in [`pkg/sql/logictest/testdata/logic_test/geospatial`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/geospatial) to call this functionality at least once. You can call `make testbaselogic FILES='geospatial' TESTFLAGS='-rewrite'` to regenerate the output. Tests here should just ensure the builtin is linked end to end (your exhaustive unit tests go the above mentioned packages!).
* Ensure the documentation is regenerated by calling `make buildshort`. You can also play with it by calling `./cockroach demo --empty` afterwards.
* Submit your PR - make sure to follow the guidelines from [creating your first PR](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR]).

You can follow #48552 for an example PR.

<sub>:robot: This issue was synced with a spreadsheet by [gsheets-to-github-issues](https://github.com/cockroachlabs/gsheet-to-github-issues) by [otan](https://github.com/otan) on 2020-06-30T05:29:57Z. Changes to titles, body and labels may be overwritten.</sub>

### PR DESCRIPTION
**Pull request by vxio**
Implement ST_Reverse on arguments {geometry}, which should adopt PostGIS behaviour.

Release note (sql change): Implemented geometry builtin `ST_Reverse`

Closes #49018

### CODE REVIEW

File: pkg/geo/geomfn/reverse.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+func reverseCoords(coords []float64, stride int) []float64 {
+	n := len(coords)
+
+	for i := 0; i < n/2; i += stride {
+		left := coords[i : i+stride]
+		right := coords[n-i-stride : n-i]
+
+		tmp := append([]float64(nil), left...)
```
Comment (otan) [inline pkg/geo/geomfn/reverse.go:68]: can we do this in place, i.e. without allocating a new array?
```
for i := 0; i < len(coords); i += stride {
   for j = 0; j < stride; j++ {
      coords[i+j], coords[len(coords)-i+j] = coords[len(coords)-i+j], coords[i+j]
    }
}
```

File: pkg/geo/geomfn/reverse.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+// reverseMulti handles reversing coordinates of MULTI* geometries with nested sub-structures
+func reverseMulti(g geom.T, ends []int) []float64 {
+	var newCoords []float64
+	flatCoords := g.FlatCoords()
+
+	prevEnd := 0
+	for _, end := range ends {
+		newCoords = append(newCoords,
+			reverseCoords(flatCoords[prevEnd:end], g.Stride())...)
```
Comment (otan) [inline pkg/geo/geomfn/reverse.go:85]: nit: one line, or one arg per line, e.g.
```
newCoords = append(
   newCoords,
   reverseCoords(flatCoords[prevEnd:end], g.Stride())...,
)

File: pkg/geo/geomfn/reverse.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+		copy(left, right)
+		copy(right, tmp)
+	}
+
+	return coords
+}
+
+// reverseMulti handles reversing coordinates of MULTI* geometries with nested sub-structures
+func reverseMulti(g geom.T, ends []int) []float64 {
+	var newCoords []float64
```
Comment (otan) [inline pkg/geo/geomfn/reverse.go:79]: i think you can specify a length here
newCoords := make([]float64, 0, len(newCoords))

you may even be able to get away with re-using the existing flat coords array:

```
prevEnd := 0
for _, end := range ends {
   copy(newCoords[prevEnd:end], reverseCoords(flatCoords[prevEnd:end]))
   prevEnd = end
}
```

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 4]
```diff
@@ -3163,6 +3163,29 @@ For flags=1, validity considers self-intersecting rings forming holes as valid a
...
+
+				ret, err := geomfn.Reverse(geometry.Geometry)
+				if err != nil {
+					return nil, err
+				}
+
+				return tree.NewDGeometry(ret), nil
+			},
+			Info: infoBuilder{
+				info: `Returns a modified geometry by reversing the order of its vertices`,
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:3184]: nit: full stop at end of info.

File: pkg/geo/geomfn/reverse.go (hunk 1 above)
Comment (vxio) [reply pkg/geo/geomfn/reverse.go]: Yep, nice suggestion. I had to tweak your code a bit though:
```go
	for i := 0; i < len(coords)/2; i += stride {
		for j := 0; j < stride; j++ {
			coords[i+j], coords[len(coords)-stride-i+j] = coords[len(coords)-stride-i+j], coords[i+j]
		}
	}

```

**Review by vxio — COMMENTED** [review]
Updated, thanks for reviewing 👍

**Review by otan — APPROVED** [review]

**Pull request merged**